ledger/participant-integration-api/src/test/lib/scala/platform/indexer/ha/TestDBLockStorageBackend.scala

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.indexer.ha
+
+import java.sql.{
+  Blob,
+  CallableStatement,
+  Clob,
+  Connection,
+  DatabaseMetaData,
+  NClob,
+  PreparedStatement,
+  SQLWarning,
+  SQLXML,
+  Savepoint,
+  Statement,
+  Struct,
+}
+import java.util.Properties
+import java.util.concurrent.Executor
+import java.{sql, util}
+
+import com.daml.platform.store.backend.DBLockStorageBackend
+
+class TestDBLockStorageBackend extends DBLockStorageBackend {
+
+  private var locks: Map[DBLockStorageBackend.Lock, Set[Connection]] = Map.empty
+
+  override def tryAcquire(
+      lockId: DBLockStorageBackend.LockId,
+      lockMode: DBLockStorageBackend.LockMode,
+  )(connection: Connection): Option[DBLockStorageBackend.Lock] = synchronized {
+    if (connection.isClosed) throw new Exception("trying to acquire on a closed connection")
+    if (!lockId.isInstanceOf[TestLockId]) throw new Exception("foreign lockId")
+    removeClosedConnectionRelatedLocks()
+    val lock = DBLockStorageBackend.Lock(lockId, lockMode)
+    def doLock(): Option[DBLockStorageBackend.Lock] = {
+      locks = locks + (lock -> (locks.getOrElse(lock, Set.empty) + connection))
+      Some(lock)
+    }
+    lockMode match {
+      case DBLockStorageBackend.LockMode.Exclusive =>
+        (
+          locks.get(lock),
+          locks.get(lock.copy(lockMode = DBLockStorageBackend.LockMode.Shared)),
+        ) match {
+          case (None, None) => doLock()
+          case (Some(connections), None) if connections == Set(connection) => doLock()
+          case _ => None // if any shared lock held, we cannot lock exclusively
```

Comment:
We should be able to upgrade from shared -> exclusive when we are the only holders, shouldn't we?

Reply:
strictly speaking yes in case of postgres. but I built the solution to avoid dependancy on lock upgrade features, so we do not have to worry about it.
pls note this test-DBLockStorageBackend is tested with the same test suite, then the real ones (Oracle, Postgres)